### PR TITLE
8338526: Don't store abstract and interface Klasses in class metaspace

### PR DESCRIPTION
**Pull request by coleenp**
This change stores InstanceKlass for interface and abstract classes in the non-class metaspace, since class metaspace will have limits on number of classes that can be represented when Lilliput changes go in.  Classes that have no instances created for them don't require compressed class pointers.  The generated LambdaForm classes are also AllStatic, and changing them to abstract moves them to non-class metaspace too.  It's not technically great to make them abstract and not final but you can't have both.  Java classfile access flags have no way of specifying something like AllStatic.

Tested with tier1-8.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526): Don't store abstract and interface Klasses in class metaspace (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [1382ced0](https://git.openjdk.org/jdk/pull/19157/files/1382ced064dec28a463d95d11bd1098fb7e15019)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19157/head:pull/19157` \
`$ git checkout pull/19157`

Update a local copy of the PR: \
`$ git checkout pull/19157` \
`$ git pull https://git.openjdk.org/jdk.git pull/19157/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19157`

View PR using the GUI difftool: \
`$ git pr show -t 19157`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19157.diff">https://git.openjdk.org/jdk/pull/19157.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19157#issuecomment-2302521420)

### CODE REVIEW

**Review by liach — COMMENTED** [review]
java.lang.invoke changes look good.

File: src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java [hunk 1]
```diff
@@ -276,7 +276,7 @@ private byte[] classFileSetup(Consumer<? super ClassBuilder> config) {
             return ClassFile.of().build(classDesc, new Consumer<>() {
                 @Override
                 public void accept(ClassBuilder clb) {
-                    clb.withFlags(ACC_FINAL | ACC_SUPER)
```
Comment (dean-long) [inline src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:279]: There's probably not much value in using ACC_FINAL here anyway.  We are only using these classes to create static methods, right?  I think ACC_INTERFACE would work here too, except for the .withSuperclass(INVOKER_SUPER_DESC) below, which maybe isn't needed.

Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:279]: Note that JVMS 4.1 requires `ACC_ABSTRACT` to be also set when `ACC_INTERFACE` is set. Also note that some classes capture class data to refer to hidden classes and method handles or lambda forms, so those fields' generation need to add `ACC_PUBLIC` flag to be usable in interfaces.

Comment (coleenp) [reply src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:279]: I feel like making it ACC_INTERFACE might cause some error if there are no public nonstatic methods, which is the case with this class. I don't know what @liach your comment means, but this code got more complicated than it was with the first version of this change.  When I talked to @rose00 he thought ACC_ABSTRACT would be okay for this.

Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:279]: Yes, you are right that we currently don't add ACC_PUBLIC flags on methods, which will fail if we add ACC_INTERFACE. Same for the fields; these LambdaForm classes use fields to store class data that's usually stored as condy, because LambdaForm is the infrastructure that condy uses (like LambdaMetafacotry cannot use lambdas)

**Comment by rkennke** [comment]
If I understand correctly, we already have limits on how many classes we can represent as compressed class-pointers. While this is nice for Lilliput, this is equally useful for non-Lilliput CCP, because addressable class-space doesn't get polluted by classes that never need to be encoded as CCP, and thus effectively increases the number of classes that we can address without resorting to -UseCompressedClassPointers.

**Comment by tstuefe** [comment]
I am surprised that this patch is so small. I would have assumed a lot of code exists that unconditionally assumes we always can encode and decode Klass*->narrowKlass.

I looked through the typical cases (eg Klass validation) and all of them seem to be okay. I will keep looking.

**Comment by coleenp** [comment]
Thanks for looking at this @tstuefe. I was pleased that the change was small but once we identify the classes as AbstractClass instead of Class in metaspace, it just falls out. I thought CDS would have more changes, but CDS is all in one space and doesn't differentiate.  It's good that the only time we compress and uncompress klass pointers is when we get them out of an object and this should keep it that way.

**Review by tstuefe — COMMENTED** [review]
Hi Coleen,

IIUC, the new "is_in_klass_space" function that now is present in all Metadata children only exists because of the template function in MetadataFactory, right? Just for the purpose of deallocation?

If so, see this proposed addition to your patch: https://gist.github.com/tstuefe/5111c735b12f6d9c3c1d32699d0820f6

This would make Metaspace::deallocate smarter - Metaspace knows whether a given pointer is in class space or not, it can do automatically the right thing. There should be no need to tell it how to deallocate that storage. (If you are worried, in debug builds there are also sanity checks).

If you do this, I think you could remove all variants of "is_in_klass_space" apart from the one in Klass.

**Comment by coleenp** [comment]
Yes, is_in_klass_space was just to direct where to deallocate the metaspace pointer.  In your patch isn't the contains metaspace call still very slow?  Or I suppose for class space, it's not because it's a fixed space.  But it's not an inlined call at all because I had to search in cpp files for the range check.
+  const bool is_class = Metaspace::contains_in_class_space(ptr);

I sort of think it might be better for the outside runtime code to control this and the metaspace call assert if its wrong.

File: src/hotspot/share/memory/allocation.hpp [hunk 2]
```diff
@@ -315,7 +315,8 @@ class MetaspaceObj {
   f(Annotations) \
   f(MethodCounters) \
   f(SharedClassPathEntry) \
-  f(RecordComponent)
+  f(RecordComponent) \
+  f(AbstractClass)
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/memory/allocation.hpp]: Yes, this is a bit inconsistent.  I like your suggestion for an improvement.  I was going to try to do this in this patch but the MetaspaceObj::Type is used for the report_metadata_oom event, so it's still needed in metaspace::allocate.  Which is unfortunate because I always get these two MetadataType and MetaspaceObj::Type things confused.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 3]
```diff
@@ -449,23 +453,24 @@ InstanceKlass* InstanceKlass::allocate_instance_klass(const ClassFileParser& par
   assert(loader_data != nullptr, "invariant");
 
   InstanceKlass* ik;
+  MetaspaceObj::Type type = (parser.is_interface() || parser.is_abstract()) ? MetaspaceObj::AbstractClassType : MetaspaceObj::ClassType;
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp]: I changed this line now (and used const for bool).

File: src/hotspot/share/oops/klass.hpp [hunk 4]
```diff
@@ -213,6 +211,8 @@ class Klass : public Metadata {
 
   virtual bool is_klass() const { return true; }
 
+  bool is_in_klass_space() const { return !is_interface() && !is_abstract(); }
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/klass.hpp]: I renamed this is_in_class_space() with the lower case 'c'.  It's still directing metaspace or indicating where the object was allocated.  Your name is a little better but I think not enough until we want to expand the things we want allocated in the class space.  As we talked about, with Tiny Class Pointers, class space will have different things in it (not that these new things need a compressed pointers).  But I think we're better off having less things in the space where their pointers can be compressed since this space is constrained.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/klass.hpp]: I have to think about this more.

**Review by coleenp — COMMENTED** [review]
Thanks for reviewing this and your comments Thomas.

**Comment by tstuefe** [comment]
> Yes, is_in_klass_space was just to direct where to deallocate the metaspace pointer. In your patch isn't the contains metaspace call still very slow? Or I suppose for class space, it's not because it's a fixed space. But it's not an inlined call at all because I had to search in cpp files for the range check.
> 
> * const bool is_class = Metaspace::contains_in_class_space(ptr);
> 
> I sort of think it might be better for the outside runtime code to control this and the metaspace call assert if its wrong.

No, I think my way is better and it will be needed anyway for TinyCP/Lilliput. We only need to do two address comparisons, that should be simple and fast. 

I opened a PR to separate the change, and in that PR I also inline the check. 

https://github.com/openjdk/jdk/pull/20701

I don't think the costs for two address comparisons matter, not with the comparatively few deallocations that happen (few hundreds or few thousand). If deallocate is hot, we are using metaspace wrong.

**Comment by tstuefe** [comment]
> I renamed this is_in_class_space() with the lower case 'c'. It's still directing metaspace or indicating where the object was allocated. Your name is a little better but I think not enough until we want to expand the things we want allocated in the class space. As we talked about, with Tiny Class Pointers, class space will have different things in it (not that these new things need a compressed pointers). But I think we're better off having less things in the space where their pointers can be compressed since this space is constrained.


How about "needs_class_space" then?

**Comment by coleenp** [comment]
> I don't think the costs for two address comparisons matter, not with the comparatively few deallocations that happen (few hundreds or few thousand). If deallocate is hot, we are using metaspace wrong.

MethodData does a lot of deallocations from metaspace because it's allocated racily.  It might be using Metaspace wrong.

**Comment by tstuefe** [comment]
> > I don't think the costs for two address comparisons matter, not with the comparatively few deallocations that happen (few hundreds or few thousand). If deallocate is hot, we are using metaspace wrong.
> 
> MethodData does a lot of deallocations from metaspace because it's allocated racily. It might be using Metaspace wrong.

I think that should be okay. This should still be an exception. I have never seen that many deallocations happening in customer cases.

File: src/hotspot/share/oops/klass.hpp [hunk 5]
```diff
@@ -201,8 +201,6 @@ class Klass : public Metadata {
   Klass(KlassKind kind);
   Klass();
 
-  void* operator new(size_t size, ClassLoaderData* loader_data, size_t word_size, TRAPS) throw();
-
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/klass.hpp:205]: ArrayKlass has its own operator new.

File: src/hotspot/share/oops/metadata.hpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2011, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (iklam) [inline src/hotspot/share/oops/metadata.hpp:2]: No more change in this file?

**Review by iklam — COMMENTED** [review]
Looks OK to me.

Maybe we should add an assert in `CompressedKlassPointers::encode_not_null()` to check that we never encode abstract and interface classes?

File: src/hotspot/share/oops/metadata.hpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/share/oops/metadata.hpp]: I removed the copyright.
I tried to add an assert to CompresssedKlassPointers::encode_not_null but CDS calls this with ill-formed Klass* pointers so the tests crash.

**Review by iklam — APPROVED** [review]

**Comment by coleenp** [comment]
Thanks Ioi.

File: src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java (hunk 1 above)
Comment (coleenp) [reply src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java:264]: Thanks for reviewing this part.

File: src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp [hunk 7]
```diff
@@ -75,13 +75,14 @@ static size_t element_size(bool compressed) {
   return compressed ? NARROW_ELEMENT_SIZE : ELEMENT_SIZE;
 }
 
-static bool can_compress_element(traceid id) {
-  return Metaspace::using_class_space() && id < uncompressed_threshold;
+static bool can_compress_element(const Klass* klass) {
+  return Metaspace::is_in_class_space(klass) &&
```
Comment (tstuefe) [inline src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp:79]:
```suggestion
  return (Metaspace::is_in_class_space(klass) || Metaspace::is_in_shared_metaspace(klass)) &&
```

Comment (coleenp) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp:79]: Is this right?  If UseCompressedClassPointers is off, then the shared metaspace isn't in compressed space?

Comment (coleenp) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: With UseCompressedClassPointers off, I think Metaspace::is_in_shared_metaspace() would still return true but I don't think he compression base is the bottom of the CDS archive.  I asked Markus to have a look.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: The code supports the JfrTraceID load barrier that enqueues tagged Klass*. It selects a more compact representation (a single word, instead of two words), if a Klass* can be compressed (i.e. there exists a compress class scheme in place (CompressedKlassPointers::encode(const_cast<Klass*>(klass)); AND the traceid (u8) value is low enough to be represented by only 4 bytes).

struct JfrEpochQueueKlassElement {
  traceid id;
  const Klass* klass;
};

struct JfrEpochQueueNarrowKlassElement {
  u4 id;
  narrowKlass compressed_klass;
};

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: // Return TRUE only if UseCompressedClassPointers is True.
  static bool using_class_space() {
    return NOT_LP64(false) LP64_ONLY(UseCompressedClassPointers);
  }
  
  I see now that was wrong for 32-bit.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: In summary, we are agnostic about which space the Klass* is located in; we only care if a valid means exists to perform an encode() and decode() operation to compress the Klass* (for 64-bit to be clear).

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: Its very rare, if at all, that an abstract or an interface would be tagged in JFR. Tags are for concrete implementations, mostly InstanceKlass*.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: I now read the JIRA issue. JFR do process loads of  java/lang/invoke/LambdaForm$MH and derivatives.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: It's fine to have each Klass* report whether it can be compressed. If not, it will be represented using the non-compressed version, which will be a bit more bloated, but no problems.

Comment (coleenp) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: narrowKlass is the result of encoding Klass* with CompressedKlassPointers::encode() which is relative to the compressed base, so if UseCompressedClassPointers is false then the encoding to narrowKlass from some other (CDS?) base isn't valid.
using_class_space() above doesn't look wrong for 32 bits.  It should return false.

With this patch, interface and abstract classes cannot be encoded and decoded to yield a valid compressed narrowKlass, since they're now allocated in the non-class metaspace.  Yes, this now a function of which space the Klass resides in.  We do crash for these classes in JFR without this change.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: I mean its wrong from JFRs perspective not to handle 32-bit outside of the call to using_class_space(), because that call will always be false for 32-bit, although the Klass* will still fit in 4-bytes.

Comment (mgronlun) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: Looks ok from a JFR perspective, Coleen.

Comment (coleenp) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: Thank you Markus.

Comment (tstuefe) [reply src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp]: Many thanks for explaining this, @mgronlun . 

@coleenp I think your code is correct insofar as it does not break anything. Your version misses out on compression opportunities for Klass in CDS, but that may not be such a big deal. You were right though that my variant was incorrect, too.

**Comment by coleenp** [comment]
@tstuefe Do you have more comments on this PR?

**Comment by tstuefe** [comment]
> @tstuefe Do you have more comments on this PR?

Sorry, I was swamped the past days. I'll take a look tomorrow.

File: src/hotspot/share/memory/metaspace.hpp [hunk 8]
```diff
@@ -159,6 +159,10 @@ class Metaspace : public AllStatic {
     return NOT_LP64(false) LP64_ONLY(UseCompressedClassPointers);
   }
 
+  static bool is_compressed_klass_ptr(const Klass* k) {
+    return using_class_space() && (is_in_class_space(k) || is_in_shared_metaspace(k));
+  }
+
```
Comment (tstuefe) [inline src/hotspot/share/memory/metaspace.hpp:165]: I propose to drop this, and instead add a utility function to `CompressedKlassPointers` like this:

```
// Returns true if p falls into the narrow Klass encoding range
inline bool CompressedKlassPointers::is_in_encoding_range(const void* p) {
  return _base != nullptr && p >= _base && p < (_base + _range);
}
```
(Probably  the `_base != nullptr` could even be left out, since `_range==0` and `_base==nullptr` for -UseCompressedClassPointers)

And then use that function in `jfrTraceIdKlass.cpp`. That file needs to use `CompressedKlassPointers` anyway because it needs to encode the Klass*.

This avoids having to rely on the exact composition of the memory regions inside the encoding range. What counts is whether the Klass pointer points into the narrow Klass encoding range.

Essentially, with CDS, memory looks like this:

```
encoding                                                               encoding              
base                                                                   end
|-----------------------------------------------------------------------|
|----CDS---| |--------------------class space---------------------------|
```

**Review by tstuefe — COMMENTED** [review]
I still worry about unforeseen implications of not every Klass having an nKlass. It will have some implications on some of my Lilliput work. I wish we could have done this earlier. 

But in any case, this seems to be a reasonable way forward. One remark inline, otherwise this looks mostly good to me.

File: src/hotspot/share/memory/metaspace.hpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/memory/metaspace.hpp:165]: oh yes that's much better.

**Review by tstuefe — APPROVED** [review]
This looks good to me.

**Comment by coleenp** [comment]
Thanks for reviewing Ioi and Thomas, and thank you Thomas for the suggested changes.
/integrate

**Comment by theRealAph** [comment]
On 9/10/24 12:42, Coleen Phillimore wrote:
 > Thanks for reviewing Ioi and Thomas, and thank you Thomas for the suggested changes.

I'm a bit concerned about this one.

I'm working on megamorphic dispatch, and a uniform representation of
compressed class pointers allows me to squeeze klass+offset into a
single 64-bit word. This in turn allows fast and simple method
lookups. I need, at least, to be able to use compressed interface
pointers. If interfaces are "somewhere else", and thus incompressible,
I can't do what I need to do. If, however, klass and non-klass
metaspaces are contiguous I guess it'd be OK, if not ideal. I'd much
rather use compressed klass pointers without having to decode them.

All I need is a way to represent interface pointers in a compact way
in lookup tables, and to be able to get from compressed class pointers
to the address. As long as interface pointers are in a 32-bit range
and there's a fast way to get from compressed class to address that's
OK.

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

**Comment by theRealAph** [comment]
On 8 Oct 2024, at 4:07, Andrew Haley wrote:

> On 9/10/24 12:42, Coleen Phillimore wrote:
>> Thanks for reviewing Ioi and Thomas, and thank you Thomas for the suggested changes.
>
> I'm a bit concerned about this one.
>
> I'm working on megamorphic dispatch, and a uniform representation of
> compressed class pointers allows me to squeeze klass+offset into a
> single 64-bit word. This in turn allows fast and simple method
> lookups. I need, at least, to be able to use compressed interface
> pointers. If interfaces are "somewhere else", and thus incompressible,
> I can't do what I need to do.

This is an interesting use of compressed klass pointers.

It looks like the compression needs to be partial, down to about 32 bits.

A motivation of this work is to compress concrete klass indexes into LESS than 32 bits, to take up less room in object layout.

So it looks like (a) having ALL klass IDs fit in 32 bits, and (b) having all CONCRETE klass IDs fit into a smaller part of the same range, would meet all requirements.  Doing it right might require TWO new companion types for Klass*, a compact concrete klass ID, and a 32-bit klass ID (supertype to compact concrete klass ID).  I think the larger one could be called narrowKlass (current name), and maybe the narrower concrete ID could be called narrowConcreteKlass or narrowestKlass or something like that.

(Because of CDS, and maybe other factors, some compact concrete klass IDs will actually point to interfaces or abstract classes.  So narrowConcreteKlass suggests a property of its referents that isn’t exactly true.)

> … If, however, klass and non-klass
> metaspaces are contiguous I guess it'd be OK, if not ideal. I'd much
> rather use compressed klass pointers without having to decode them.

The way I read this is to think about putting, not klass and non-klass, but concrete-klass and non-concrete-klass IDs in the same range.  (By ID I mean a compressed Klass pointer, or alternatively an index into some sort of special table that we haven’t needed to invent.)

> All I need is a way to represent interface pointers in a compact way
> in lookup tables, and to be able to get from compressed class pointers
> to the address. As long as interface pointers are in a 32-bit range
> and there's a fast way to get from compressed class to address that's
> OK.

**Comment by merykitty** [comment]
@theRealAph I do not know your idea regarding this matter but is compressing interface pointers independently from concrete class pointers help?

**Comment by rose00** [comment]
If the interfaces had a compact numbering, and there were a side table mapping the compact numbers to interface Klass* pointers, then I think Andrew's code would still work (with natural adjustments).

But managing such a side table is at least as complicated as just doing the pointer compression stuff we do.  (Hence my comment that we haven't had to invent side tables yet.)

Because of CDS I don't think we can treat concretes and abstracts (or even just classes and interfaces) as disjoint metadata types with separate independent compression tactics or representations.  I think we need a subtype/supertype relation between the "narrowest" and merely "narrower" klass IDs.

**Comment by rkennke** [comment]
AFAIK, @tstuefe (currently on vacation) has a working prototype of a Klass-lookup-table with good performance and reasonable ‘management cost’. This would make make many things much simpler and also help solve this problem because it makes irrelevant where a Klass lives.

Roman

**Comment by theRealAph** [comment]
On 10/9/24 06:53, John Rose wrote:

 > It looks like the compression needs to be partial, down to about 32 bits.

Yes, that's true. Slightly fewer bits would be nice, for tags.

 > So it looks like (a) having ALL klass IDs fit in 32 bits, and (b)
 > having all CONCRETE klass IDs fit into a smaller part of the same
 > range, would meet all requirements. Doing it right might require TWO
 > new companion types for Klass*, a compact concrete klass ID, and a
 > 32-bit klass ID (supertype to compact concrete klass ID). I think
 > the larger one could be called narrowKlass (current name), and maybe
 > the narrower concrete ID could be called narrowConcreteKlass or
 > narrowestKlass or something like that.

I see, I think.

Simply keeping all of metadata in a single 32-bit range would be very
useful. We could have instantiable Klasses in one region, followed by
non-instantiable Klasses.

 > (Because of CDS, and maybe other factors, some compact concrete
 > klass IDs will actually point to interfaces or abstract classes. So
 > narrowConcreteKlass suggests a property of its referents that isn’t
 > exactly true.)

 > The way I read this is to think about putting, not klass and
 > non-klass, but concrete-klass and non-concrete-klass IDs in the same
 > range. (By ID I mean a compressed Klass pointer, or alternatively an
 > index into some sort of special table that we haven’t needed to
 > invent.)

While I've now thought of a way to encode Klass pointers that doesn't
need them all to be in the same 32-bit range, it may well be very
useful in other contexts to ensure that they will be.

Maybe we could reduce the footprint of Klass instances. Right now,
though, my invokeinterface prototypes *increase* the size of Klass
instances, so I really want to have the possibility of using
compressed metadata pointers.

For what it's worth, I'm relieved that we're not yet making Klass
pointers a table index. Another chained load in the path of method
dispatch, at a time when I'm trying to get rid of chained loads, would
be a Bad Thing for me.

-- 
Andrew Haley  (he/him)
Java Platform Lead Engineer
Red Hat UK Ltd. <https://www.redhat.com>
https://keybase.io/andrewhaley
EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

**Comment by theRealAph** [comment]
> @theRealAph I do not know your idea regarding this matter but does compressing interface pointers independently from concrete class pointers help?

Of course that is possible, but it add complexity; and my goal is reducing complexity.

**Comment by theRealAph** [comment]
> If the interfaces had a compact numbering, and there were a side table mapping the compact numbers to interface Klass* pointers, then I think Andrew's code would still work (with natural adjustments).

Probably, yes. I can pack klass ID+ [iv]table offset into a word to identify a method, or use a compressed metadata pointer. But I have a data dependency on a table load to get from a compressed Klass ID of some sort to a `Klass*`, that's a 3-to-5-cycle stall.

> But managing such a side table is at least as complicated as just doing the pointer compression stuff we do. (Hence my comment that we haven't had to invent side tables yet.)
> 
> Because of CDS I don't think we can treat concretes and abstracts (or even just classes and interfaces) as disjoint metadata types with separate independent compression tactics or representations. I think we need a subtype/supertype relation between the "narrowest" and merely "narrower" klass IDs.

I think that's right.

Notwithstanding any side tables etc., it would be nice to make sure that all metadata is reachable with a 4Gbyte offset from some base. It's not so hard to make sure that concrete and abstract `Klass*` (and all `Method*`s) can be in that range, simply(?) by allocating everything contiguously.

**Comment by rose00** [comment]
> simply(?) by allocating everything contiguously

Suddenly I seem to hear Boromir and Yoda, in unison, saying, "One does not simply."

**Comment by theRealAph** [comment]
> AFAIK, @tstuefe (currently on vacation) has a working prototype of a Klass-lookup-table with good performance and reasonable ‘management cost’. This would make make many things much simpler and also help solve this problem because it makes irrelevant where a Klass lives.

Yeah, I get that, and I'm sure @tstuefe has done a great job. But it goes to the fundamental theorem of software engineering (FTSE), Wheeler's statement that "We can solve any problem by introducing an extra level of indirection," often followed by "…except for the problem of too many levels of indirection."

**Comment by rose00** [comment]
Sometimes, if you are very clever and determined, you can do the very non-simple thing of putting some items at the least level of indirection, and other items at further levels of indirection.  You'd try to put stuff you need frequently closer to the root address.

Currently everything is equally close to the root address, except that IIRC we try to put a few really performance sensitive things on the earliest cache lines.

Splitting one Klass structure into a near and a far part is doable in principle, but the complexity adds up quickly, and there is also a problem deciding what to put in the near part.

Specifically, you probably want some v-table entries, but not all v-table entries, in the near part.  (Why not all?  Because near space is relatively scarce.  The near part might even be fixed in size, depending on design.)

The bottom line for me:  I like flatter data and shorter indirection chains.  The main exception is that I ALSO like to separate hot from cold data, when the opportunity arises, and cold data is often at the end of an unused indirection.

(I think there is relatively little cold data in metaspace, and it is probably already packed inside a pointer to an Array.)
